src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs

```diff
@@ -22,63 +21,59 @@ public FeatureShapeDescriptor(IFeatureInfo feature, string shapeType)
+        private readonly List<FeatureShapeDescriptor> _alternationDescriptors = [];
+        private readonly List<string> _wrappers = [];
+        private readonly List<string> _bindingSources = [];
+        private readonly Dictionary<string, ShapeBinding> _bindings = new(StringComparer.OrdinalIgnoreCase);
+        private readonly List<Func<ShapeCreatingContext, Task>> _creatingAsync = [];
+        private readonly List<Func<ShapeCreatedContext, Task>> _createdAsync = [];
+        private readonly List<Func<ShapeDisplayContext, Task>> _displayingAsync = [];
+        private readonly List<Func<ShapeDisplayContext, Task>> _processingAsync = [];
+        private readonly List<Func<ShapeDisplayContext, Task>> _displayedAsync = [];
 
         public ShapeDescriptorIndex(
             string shapeType,
             IEnumerable<string> alterationKeys,
             ConcurrentDictionary<string, FeatureShapeDescriptor> descriptors)
         {
             ShapeType = shapeType;
-            _descriptors = descriptors;
-
-            // pre-calculate as much as we can
-            _alternationDescriptors = alterationKeys
-                .Select(key => _descriptors[key])
-                .ToList();
-
-            _wrappers = _alternationDescriptors
-                .SelectMany(sd => sd.Wrappers)
-                .ToList();
-
-            _bindingSources = _alternationDescriptors
-                .SelectMany(sd => sd.BindingSources)
-                .ToList();
-
-            _bindings = _alternationDescriptors
-                .SelectMany(sd => sd.Bindings)
-                .GroupBy(kv => kv.Key, StringComparer.OrdinalIgnoreCase)
-                .Select(kv => kv.Last())
-                .ToDictionary(kv => kv.Key, kv => kv.Value, StringComparer.OrdinalIgnoreCase);
-
-            _creatingAsync = _alternationDescriptors
-                .SelectMany(sd => sd.CreatingAsync)
-                .ToList();
-
-            _createdAsync = _alternationDescriptors
-                .SelectMany(sd => sd.CreatedAsync)
-                .ToList();
-
-            _displayingAsync = _alternationDescriptors
-                .SelectMany(sd => sd.DisplayingAsync)
-                .ToList();
-
-            _processingAsync = _alternationDescriptors
-                .SelectMany(sd => sd.ProcessingAsync)
-                .ToList();
-
-            _displayedAsync = _alternationDescriptors
-                .SelectMany(sd => sd.DisplayedAsync)
-                .ToList();
+
+            // Pre-calculate as much as we can for performance reasons.
+            foreach (var alterationKey in alterationKeys)
+            {
+                if (!descriptors.TryGetValue(alterationKey, out var alternationDescriptor))
+                {
+                    continue;
+                }
+
+                _alternationDescriptors.Add(alternationDescriptor);
+
+                if (alternationDescriptor.Wrappers.Count > 0)
```

Comment:
`AddRange() [does this check](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L253) too.